modules/core/shared/src/main/scala/io/circe/Decoder.scala

```diff
@@ -87,15 +87,17 @@ trait Decoder[A] extends Serializable { self =>
   }
 
   private[this] def cursorToDecodingFailure(cursor: ACursor) = {
-    val history = cursor.history
-    val historyToFailedCursor = history.takeWhile(_ != CursorOp.DeleteGoParent)
-    val field = CursorOp.opsToPath(historyToFailedCursor).replaceFirst("^\\.", "")
-    val down = cursor.downField(field)
-    if (down.succeeded) {
-      DecodingFailure(s"Couldn't decode $field", history)
-    } else {
-      DecodingFailure(MissingField, history)
-    }
+    val reason: DecodingFailure.Reason =
+      cursor match {
+        case cursor: FailedCursor =>
+          if (cursor.missingField) {
+            DecodingFailure.Reason.MissingField
+          } else {
+            DecodingFailure.Reason.CustomReason("Unable to decode JSON")
```

Comment:
This change in reason for the non-missing field case could cause issues... just like when we changed them for 0.14.2.   Would the performance issue we are trying to solve be that bad if we keep this message the same?

Reply:
I know we made a bit of a headache for some users, so wondering what can be done about that... hate to thrash too much 😅